bail out of event blocking for Adobe CEP bug (fix #10366)

ericdrobinson opened the pull request:
fix #10366

<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
Some [Adobe CEP](https://github.com/Adobe-CEP) [environments](https://github.com/Adobe-CEP/CEP-Resources/blob/master/CEP_9.x/Documentation/CEP%209.0%20HTML%20Extension%20Cookbook.md#applications-integrated-with-cep) have a broken `Event.timeStamp` implementation that breaks event blocking logic. The issue specifically affects host applications running on macOS with CEP version 9.3 and below. This workaround is OS-agnostic so as to maintain identical behavior across platforms. More details can be found in issue #10366.

ericdrobinson committed:
fix: bail out of event blocking for Adobe CEP bug

Some Adobe CEP environments have a broken Event.timeStamp implementation that breaks event blocking
logic. The issue specifically affects host applications running on macOS with CEP version 9.3 and
below. This workaround is OS-agnostic so as to maintain identical behavior across platforms. More
details can be found in issue #10366.

fix #10366

derrickb commented:
One note:
`var isCEP = inBrowser && window.__adobe_cep__ !== 'undefined';`
I believe you're missing `typeof`, or change `'undefined'` to `undefined`.

derrickb commented on src/core/util/env.js:
```diff
@@ -16,6 +16,7 @@ export const isIOS = (UA && /iphone|ipad|ipod|ios/.test(UA)) || (weexPlatform ==
 export const isChrome = UA && /chrome\/\d+/.test(UA) && !isEdge
 export const isPhantomJS = UA && /phantomjs/.test(UA)
 export const isFF = UA && UA.match(/firefox\/(\d+)/)
+export const isCEP = inBrowser && window.__adobe_cep__ !== 'undefined'
```
`undefined` shouldn't be a string here.

ericdrobinson replied:
Good catch! I'll get that cleaned up!

ericdrobinson commented:
@derrickb Fixed! Thanks for catching that!! 😅

posva commented on src/core/util/env.js:
```diff
@@ -16,6 +16,7 @@ export const isIOS = (UA && /iphone|ipad|ipod|ios/.test(UA)) || (weexPlatform ==
 export const isChrome = UA && /chrome\/\d+/.test(UA) && !isEdge
 export const isPhantomJS = UA && /phantomjs/.test(UA)
 export const isFF = UA && UA.match(/firefox\/(\d+)/)
+export const isCEP = inBrowser && window.__adobe_cep__ !== undefined
```
I think checking `window.__adobe_cep__` is enough since we know `window` is defined

ericdrobinson replied:
Can you clarify? Which of the following do you mean:

1. ```js
    export const isCEP = window.__adobe_cep__ !== undefined
    ```
2. ```js
    export const isCEP = inBrowser && window.__adobe_cep__
    ```

The first would cause issues where `window` is not defined (i.e. not in `inBrowser` contexts) while the second seems a bit ambiguous and doesn't follow the `undefined` check style of other options.

Or did you mean something  else?

ericdrobinson replied:
I spoke with @posva in a separate forum (Discord direct chat) and it was determined that using the `isObject` utility function was the best approach to validating `window.__adobe_cep__`.

posva commented on src/platforms/web/runtime/modules/events.js:
```diff
@@ -44,6 +44,14 @@ function createOnceHandler (event, handler, capture) {
+// issue is restricted to macOS, the fix is OS-agnostic to keep behavioral
+// differences to a minimum.
+const isCEP93orEarlier = isCEP && ((maxBadMajor, maxBadMinor) => {
+  const version = JSON.parse(window.__adobe_cep__.getCurrentApiVersion())
```
are we sure this function exist in **all** versions of adobe_cep?

posva replied:
If it exist in all version till today, I think we should check if the function exists and if it doesn't return false to prevent breaking changes with adobe_cep api in the future

derrickb replied:
It has existed in all CEP versions since its inception in [version 4.x ](https://github.com/Adobe-CEP/CEP-Resources/blob/5880e20b7cb2dc09e6b226f15a33165a08c0d831/CEP_4.x/CSInterface.js#L812)

To my knowledge, Adobe has never removed functions from their CEP API, even deprecated ones. However, a check for the function is still a good idea.

ericdrobinson replied:
>are we sure this function exist in all versions of adobe_cep?

Not 100%, no. The table shows that the earliest version of  **HTML**-based CEP was "CEP 4". Adobe has an extraordinarily bad habit of not publishing details of the CEP minor versions for their apps so it is entirely possible that "version 4" means "version 4.2." That said, the opposite is also possible.

>If it exist in all version till today, I think we should check if the function exists and if it doesn't return false to prevent breaking changes with adobe_cep api in the future

@bbb999 might be able to shed more light on this but my understanding is that the upcoming CEP 9.4.x is likely to be the last version of CEP. Adobe is _beginning_ to transition their apps to a new technology (known as UXP) and are starting to phase out CEP. There may be bugfix releases but I'd be willing to bet that we won't see any core API changes before CEP's final death knell.

>However, a check for the function is still a good idea.

Agreed. I can make that change for sure.

bbb999 replied:
Adobe has made no announcements regarding [above], but Eric's speculation is entirely plausible. :)

ericdrobinson replied:
@bbb999 I'm going to add a check for the existence of the `getCurrentApiVersion` function. It's non-existence, however, can mean one of two things:

1. This is an archaic version of CEP that pre-dates that function.
2. This is a future version of CEP that redefined that function as something else.

Given the trajectory of CEP over its lifetime, it seems that option 2 is far, far less likely than option 1. Would it be safe to assume, then, that if:

```js
typeof __adobe_cep__.getCurrentApiVersion !== 'function'
```

then we're more likely to be working with an archaic version of CEP than an as-yet-unreleased one?

ericdrobinson replied:
Spoke with @bbb999 in another forum (Adobe Devs community Slack) and the response was:

>[Option 1], definitely.

ericdrobinson committed:
fix: Adjust existence check to be more explicit

Improve type checking of the `__adobe_cep__` object. It shouldn't just be `undefined` but explicitly
an object.

ericdrobinson committed:
fix: Validate function existence prior to invoking

It is possible that extremely early versions of Adobe CEP did not have the `getCurrentApiVersion`
function defined. In such cases, assume "CEP 9.3 or earlier".

ericdrobinson commented:
@posva Fixes applied.  Please let me know if there is anything further!

ericdrobinson commented:
For the record, [the decision](https://github.com/vuejs/vue/pull/10459#discussion_r322929456) to assume that:

   >`__adobe_cep__.getCurrentApiVersion === undefined` => super old CEP version

was correct.

I just ran across this statement in [the CEP Cookbook](https://github.com/Adobe-CEP/CEP-Resources/blob/master/CEP_9.x/Documentation/CEP%209.0%20HTML%20Extension%20Cookbook.md#api-version):

>Please note [that the `getCurrentApiVersion`] API itself is available only since 4.2.0. If you get an error saying `getCurrentApiVersion` is undefined, then you are running in CEP 4.0 or 4.1. Otherwise, the value returned will tell you the version of CEP integrated by the Adobe product.

Awesome. 💪
